projects/hathor-wallet-headless/hsm-integration.md

```diff
@@ -0,0 +1,313 @@
+
+## Headless implementation
+[headless-implementation]: #headless-implementation
+The headless application will be responsible for calling the *HSM Client* directly.
+
+The connection authentication credentials will be stored securely on the environment variables, allowing the leverage of other security best practices to avoid credential leaks.
+```sh
+# On env variables
+HSM_DEVICE_HOST="192.168.0.1"
+HSM_DEVICE_USERNAME="my_username"
+HSM_DEVICE_PASSWORD="my_password"
+```
+
+A configuration attribute will be added to the `config.js` file, similar to the existing `seeds` attribute, mapping the headless `walletId`s to HSM object key names.
+```js
+// On config.js
+const hsmWallets = {
+  'headless-wallet-id': 'HSM_KEY_NAME'
+}
+```
+
+When requesting to `[POST] /start` a wallet with the `walletId` within this mapped list, the headless considers this to be a hardware wallet and realizes the first request to the HSM to validate the existence of the key name.
+
+> **Possibility 1: The key is available**<br>
+> The case flow continues
+>
+> **Possibility 2: The key is not available**<br>
+> An error is thrown indicating the HSM has not been properly configured to have the `xPriv` on this specific key.
+
+The key being available, the next call requests the wallet's `xPub`. The parsing of `xPub` information received from the HSM can be based on the desktop hardware wallet [`handlePublicKeyData` function](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/screens/StartHardwareWallet.js#L106-L138).
+
+If the generated `xPub` is not valid, an error is thrown to the user indicating that the designated HSM key does not contain a valid `BIP32 xPriv` and is unsuitable for using as a wallet.
+
+Once the wallet is started successfully, the signature of HSM inputs can be made through a `Tx Proposal`, using the Atomic Swap API. The `wallet/atomic-swap/get-my-signatures` route will be modified to identify that the wallet is related to an HSM and will send the data to be signed to the *HSM Client*. This can be based on the desktop hardware wallet [`ledger/sendTx()` function](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/utils/ledger.js#L126-L166).
+
+_Note:_ For this initial version, only this route will be adapted for using the hardware wallet, since it's the only one that has an implementation close enough to the Desktop Ledger. The other endpoints, such as `wallet/simple_send_tx` will require more complex refactorings and will be discussed on the _Future Possibilites_ section.
+
+| ⚠ Incomplete PoC - sign tx                                                                           |
+|:-----------------------------------------------------------------------------------------------------|
+| This step was also not fully implemented on the PoC and can add uncertainty to the development phase |
```

Comment:
I think we should remove this from warning from the design

Reply:
Removed.